platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+		result_local.error_code = ML_LIB_FAILED;
+		goto release_all_tensors;
+	}
+
+	/* Verify output tensors and store them to output */
+	if (output_tensors_to_data(output_tensors, ml_info->num_outputs, param,
+				   output_info, data, &result_local)) {
+		_ODP_ERR("Output tensors to data failed\n");
+		goto release_all_tensors;
+	}
+
+	retval = 1;
+
+release_all_tensors:
+	for (uint32_t i = 0; i < ml_info->num_outputs; i++)
+		ort_api->ReleaseValue(output_tensors[i]);
+
+release_input_tensors:
+	for (uint32_t i = 0; i < ml_info->num_inputs; i++)
+		ort_api->ReleaseValue(input_tensor[i]);
+
+init_fail:
+	if (param && param->result)
+		*param->result = result_local;
+
+	odp_ticketlock_lock(&mdl->lock);
+	mdl->state = ML_STATE_LOADED;
+	odp_ticketlock_unlock(&mdl->lock);
+
+	return retval;
+}
+
+int odp_ml_run_multi(odp_ml_model_t model, const odp_ml_data_t data[],
+		     const odp_ml_run_param_t param[], int num)
+{
+	int i;
+	int ret;
+
+	if (odp_unlikely(num < 1)) {
+		_ODP_ERR("Bad number of runs\n");
+		return -1;
+	}
+
+	for (i = 0; i < num; i++) {
+		if (param)
+			ret = odp_ml_run(model, &data[i], &param[i]);
+		else
+			ret = odp_ml_run(model, &data[i], NULL);
+
+		if (odp_unlikely(ret != 1))
+			break;
+	}
+
+	if (odp_unlikely(i == 0))
+		return ret;
+
+	return i;
+}
+
+int odp_ml_run_start(odp_ml_model_t model, const odp_ml_data_t *data,
+		     const odp_ml_compl_param_t *compl_param,
+		     const odp_ml_run_param_t *run_param)
+{
+	int ret;
+	ml_model_t *mdl = ml_model_from_handle(model);
+
+	if (odp_unlikely(model == ODP_ML_MODEL_INVALID)) {
+		_ODP_ERR("Bad model handle\n");
+		return -1;
+	}
+
+	if (odp_unlikely(!compl_param)) {
+		_ODP_ERR("Completion parameter is NULL\n");
+		return -1;
+	}
+
+	/* Check completion mode */
+	if (odp_unlikely(check_compl_param(compl_param, mdl->max_compl_id, false))) {
+		_ODP_ERR("Bad ML job completion parameter\n");
+		return -1;
+	}
+
+	if (compl_param->mode == ODP_ML_COMPL_MODE_POLL)
+		odp_atomic_store_rel_u32(&mdl->compl_status[compl_param->compl_id], 0);
+
+	ret = odp_ml_run(model, data, run_param);
+
+	if (odp_unlikely(ret < 1))
+		return ret;
+
+	/* Send a completion event to the given queue */
+	if (compl_param->mode == ODP_ML_COMPL_MODE_EVENT) {
+		odp_ml_run_result_t *result;
+		odp_buffer_t buf = (odp_buffer_t)(uintptr_t)compl_param->event;
+
+		_odp_buffer_subtype_set(buf, ODP_EVENT_ML_COMPL_RUN);
+
+		result = odp_buffer_addr(buf);
+		result->error_code = 0;
+		result->user_ptr = compl_param->user_ptr;
+
+		if (odp_unlikely(odp_queue_enq(compl_param->queue, compl_param->event))) {
+			_ODP_ERR("Completion event enqueue failed %" PRIu64 "\n",
+				 odp_queue_to_u64(compl_param->queue));
+			return -1;
+		}
+
+		return 1;
+	}
+
+	/* compl_param->mode == ODP_ML_COMPL_MODE_POLL */
+	odp_atomic_store_rel_u32(&mdl->compl_status[compl_param->compl_id], 1);
+
+	/* odp_ml_model_run() succeeded */
+	mdl->result[compl_param->compl_id].user_ptr = compl_param->user_ptr;
```

Comment:
Comment can be removed, it refers to wrong function name anyway. user_ptr write should happen before status store-release.